README.md
# Sysrandom [![Gem Version][gem-image]][gem-link] [![Build Status][build-image]][build-link] [![ISC licensed][license-image]][license-link]

[gem-image]: https://badge.fury.io/rb/sysrandom.svg
[gem-link]: https://rubygems.org/gems/sysrandom
[build-image]: https://secure.travis-ci.org/cryptosphere/sysrandom.svg?branch=master
[build-link]: https://travis-ci.org/cryptosphere/sysrandom
[license-image]: https://img.shields.io/badge/license-ISC-blue.svg
[license-link]: https://github.com/cryptosphere/sysrandom/blob/master/LICENSE.txt

Secure random number generation for Ruby using system RNG facilities e.g. `/dev/urandom`, `getrandom(2)`

## Why?

System/OS-level random number generators like `/dev/urandom` and `getrandom(2)`
provide the best option for generating cryptographically secure random numbers.

Ruby's built-in SecureRandom does not provide this, but instead uses OpenSSL's
userspace RNG. This has been a [source of vulnerabilities][emboss] in Ruby, and
an [open Ruby bug ticket][bug] contains much discussion on the issue with no
clear path to resolution.

This gem aims to solve the problem.

## Description

In cryptography circles, [the prevailing advice is to use OS RNG functionality][/dev/urandom],
namely `/dev/urandom` or equivalent calls which use an OS-level CSPRNG to
produce random numbers.

This gem provides an easy-to-install repackaging of the `randombytes`
functionality from [libsodium] for the purpose of generating secure random
numbers trustworthy for use in cryptographic contexts, such as generating
cryptographic keys, initialization vectors, or nonces.

The following random number generators are utilized:

| Platform | RNG                                                    |
|----------|--------------------------------------------------------|
| Linux    | [getrandom(2)] if available, otherwise [/dev/urandom]  |
| Windows  | [RtlGenRandom] CryptGenRandom without CryptoAPI deps   |
| OpenBSD  | [arc4random(3)] with ChaCha20 CSPRNG (not RC4)         |
| JRuby    | [NativePRNGNonBlocking] on Java 8, otherwise SHA1PRNG  |
| Others   | [/dev/urandom]                                         |

[emboss]:        https://emboss.github.io/blog/2013/08/21/openssl-prng-is-not-really-fork-safe/
[bug]:           https://bugs.ruby-lang.org/issues/9569
[libsodium]:     https://github.com/jedisct1/libsodium
[getrandom(2)]:  http://man7.org/linux/man-pages/man2/getrandom.2.html
[/dev/urandom]:  http://sockpuppet.org/blog/2014/02/25/safely-generate-random-numbers/
[RtlGenRandom]:  https://msdn.microsoft.com/en-us/library/windows/desktop/aa387694(v=vs.85).aspx
[arc4random(3)]: http://man.openbsd.org/arc4random.3
[NativePRNGNonBlocking]: https://tersesystems.com/2015/12/17/the-right-way-to-use-securerandom/

## Supported Platforms

Sysrandom is tested on the following Ruby implementations:

* Ruby (MRI) 2.0, 2.1, 2.2, 2.3
* JRuby 9.1.1.0

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'sysrandom'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install sysrandom

## Usage

`Sysrandom` aims to be API-compatible with Ruby's built-in `SecureRandom` class,
but always prefers OS-level RNG wherever it's available:

```ruby
>> Sysrandom.random_number(42)
=> 15
>> Sysrandom.random_bytes(32)
=> "\xD6J\xB3\xD2\x8B\x7F*9D\xB7\xF9\xEA\xE2\\\xAAH\tV#\xEC\x84\xE3E\r\x97\xB9\b\xFCH\x17\xA0\v"
>> Sysrandom.base64(32)
=> "WXPkxfAuLRpnI6Z4zFb4E+MIenx6w6vKhe01+rMPuIQ="
>> Sysrandom.urlsafe_base64(32)
=> "37rsMfR4X8g7Bb-uDJEekRHnB3r_7nO03cv52ilaWqE="
>> Sysrandom.hex(32)
=> "c950496ce200abf7d18eb1414e9206c6335f971a37d0394114f56439b59831ba"
>> Sysrandom.uuid
=> "391c6f52-8017-4838-9790-131a9b979c63"
```

* [SecureRandom API docs](http://ruby-doc.org/stdlib-2.0.0/libdoc/securerandom/rdoc/SecureRandom.html)

## Patching SecureRandom with Sysrandom

Since Sysrandom is SecureRandom-compatible, it can be patched in-place of
SecureRandom if you prefer its RNG behavior.

To do this, require `sysrandom/securerandom`:

```ruby
>> SecureRandom
=> SecureRandom
>> require "sysrandom/securerandom"
=> true
>> SecureRandom
=> Sysrandom
>> SecureRandom.hex(32)
=> "d1bbe8c1ab78fc2fe514c5623d913a27ffd2dcdc9e002f3b358bb01a996962f1"
```

## Contributing

* Fork this repository on Github
* Make your changes and send a pull request
* If your changes look good, we'll merge them

## Copyright

Copyright (c) 2013-2016 Frank Denis, Tony Arcieri. See LICENSE.txt for further details.


CHANGES.md
## 1.0.4 (2016-12-04)

* [#15](https://github.com/cryptosphere/sysrandom/pull/16)
  Argument handling fixups.
  ([@tarcieri])

## 1.0.3 (2016-09-27)

* [#14](https://github.com/cryptosphere/sysrandom/pull/14)
  Return empty string on random_bytes(0) as SecureRandom does.
  ([@grempe])

## 1.0.2 (2016-06-06)

* [#12](https://github.com/cryptosphere/sysrandom/pull/12)
  Remove unnecessary SecureRandom reference in 'hex' method.
  ([@tarcieri])

* [#6](https://github.com/cryptosphere/sysrandom/pull/6)
  Raise ArgumentError if 0 random bytes are requested (new minimum is 1).
  ([@azet])

## 1.0.1 (2016-05-29)

* [#11](https://github.com/cryptosphere/sysrandom/pull/11)
  Use NativePRNGNonBlocking for JRuby if available.
  ([@coda])

* [#8](https://github.com/cryptosphere/sysrandom/pull/8)
  Upstream libsodium change: wait for `/dev/random` to be seeded before reading from `/dev/urandom`.
  ([@tarcieri])


## 1.0.0 (2016-05-28)

* Initial release.


[@tarcieri]: https://github.com/tarcieri
[@coda]: https://github.com/coda
[@azet]: https://github.com/azet
[@grempe]: https://github.com/grempe


spec/spec_helper.rb
require "sysrandom"

RSpec.configure(&:disable_monkey_patching!)


spec/sysrandom_spec.rb
require "spec_helper"

RSpec.describe Sysrandom do
  it "has a version number" do
    expect(Sysrandom::VERSION).not_to be nil
  end

  describe ".random_number" do
    it "creates floats between 0..1 if no argument is given" do
      n = described_class.random_number
      expect(n).to be_a Float
      expect(n).to be > 0
      expect(n).to be < 1
    end

    it "returns a Fixnum if given a Fixnum as an argument" do
      expect(described_class.random_number(42)).to be_a Fixnum
    end

    it "returns a Float if given a Float as an argument" do
      expect(described_class.random_number(42.0)).to be_a Float
    end
  end

  describe ".random_bytes" do
    it "creates strings with random bytes" do
      string = described_class.random_bytes
      expect(string).to be_a String

      string2 = described_class.random_bytes
      expect(string).to_not eq string2 # randomness is hard to test for
    end

    it "honors its length argument" do
      expect(described_class.random_bytes(42).size).to eq 42
    end

    it "allows nil as a length argument" do
      expect(described_class.random_bytes(nil).size).to eq 16
    end

    # SecureRandom's wacky default string size
    it "creates strings of length 16 by default" do
      expect(described_class.random_bytes.size).to eq 16
    end

    it "returns an empty string if zero bytes requested" do
      expect(described_class.random_bytes(0)).to eq ""
    end

    it "generates a random binary string of specified length" do
      (1..64).each do |idx|
        bytes = described_class.random_bytes(idx)
        expect(bytes).to be_a String
        expect(bytes.size).to eq idx
      end

      expect(described_class.random_bytes(2.2).length).to eq(2)
    end

    it "generates different binary strings with subsequent invocations" do
      # quick and dirty check, but good enough
      values = []
      256.times do
        val = described_class.random_bytes
        # make sure the random bytes are not repeating
        expect(values.include?(val)).to be(false)
        values << val
      end
    end

    it "raises ArgumentError on negative arguments" do
      expect { described_class.random_bytes(-1) }.to raise_error(ArgumentError, "negative string size")
    end

    it "calls to_int on the number passed as an arg" do
      # jruby doesn't do this test
      unless defined? JRUBY_VERSION
        obj = double("to_int")
        expect(obj).to receive(:to_int) { 5 }
        expect(described_class.random_bytes(obj).size).to eq(5)
      end
    end
  end

  describe ".base64" do
    it "creates random base64 strings" do
      base64 = described_class.base64(16)
      expect(Base64.decode64(base64)).to be_a String
    end
  end

  describe ".hex" do
    it "creates random hexadecimal strings" do
      hex = described_class.hex(16)
      expect([hex].pack("h*")).to be_a String
    end
  end

  describe ".urlsafe_base64" do
    it "creates unpadded strings by default" do
      expect(described_class.urlsafe_base64).not_to include("=")
    end

    it "optionally creates padded strings" do
      expect(described_class.urlsafe_base64(nil, true)).to include("=")
    end

    it "creates valid urlsafe_base64 strings" do
      base64 = described_class.urlsafe_base64(nil, true)
      expect(Base64.urlsafe_decode64(base64)).to be_a String
    end
  end

  describe ".uuid" do
    it "creates random UUIDs" do
      uuid = described_class.uuid
      expect(uuid).to match(/[a-f0-9]{8}-[a-f0-9]{4}-4[a-f0-9]{3}-[89ab][a-f0-9]{3}-[a-f0-9]{12}/)
    end
  end
end


spec/sysrandom/securerandom_spec.rb
require "spec_helper"

RSpec.describe "sysrandom/securerandom" do
  it "patches SecureRandom to be Sysrandom when loaded" do
    require "securerandom"
    expect(SecureRandom).not_to eq Sysrandom

    require "sysrandom/securerandom"
    expect(SecureRandom).to eq Sysrandom
  end
end


ext/sysrandom/sysrandom.c
#include "ruby.h"

/* How many bytes to generate by default, matching SecureRandom
 * Why SecureRandom has a default size argument is beyond me */
#define DEFAULT_N_BYTES 16

static VALUE mSysrandom = Qnil;

static VALUE Sysrandom_random_bytes(int argc, VALUE *argv, VALUE self);

/* From randombytes_sysrandom.c */
void __randombytes_sysrandom_buf(void * const buf, const size_t size);

void Init_sysrandom()
{
    mSysrandom = rb_define_module("Sysrandom");

    rb_define_singleton_method(mSysrandom, "random_bytes", Sysrandom_random_bytes, -1);
}

/**
 *  call-seq:
 *    Sysrandom#random_bytes(n=nil) -> String
 *
 * ::random_bytes generates a random binary string via the OS CSPRNG.
 *
 * The argument n specifies how long the resulting string should be.
 *
 * For compatibility with SecureRandom, if n is not specified, 16 is assumed.
 *
 * The resulting string may contain any byte (i.e. "x00" - "xff")
 *
 */
static VALUE
Sysrandom_random_bytes(int argc, VALUE * argv, VALUE self)
{
    VALUE n_obj, str;
    int n;

    if (rb_scan_args(argc, argv, "01", &n_obj) == 1) {
        n = NUM2INT(n_obj);

        if(n < 0) {
          rb_raise(rb_eArgError, "negative string size");
        }
    } else {
        n = DEFAULT_N_BYTES;
    }

    if(n > 0) {
        str = rb_str_new(0, n);
        __randombytes_sysrandom_buf(RSTRING_PTR(str), n);
    } else {
        str = rb_str_new2("");
    }

    return str;
}


ext/sysrandom/sysrandom_ext.c
#include "ruby.h"

/* How many bytes to generate by default, matching SecureRandom
 * Why SecureRandom has a default size argument is beyond me */
#define DEFAULT_N_BYTES 16

static VALUE mSysrandom = Qnil;

static VALUE Sysrandom_random_uint32(VALUE self);
static VALUE Sysrandom_random_bytes(int argc, VALUE *argv, VALUE self);

/* From randombytes_sysrandom.c */
uint32_t __randombytes_sysrandom(void);
void __randombytes_sysrandom_buf(void * const buf, const size_t size);

void Init_sysrandom_ext()
{
    mSysrandom = rb_define_module("Sysrandom");

    rb_define_singleton_method(mSysrandom, "__random_uint32", Sysrandom_random_uint32, 0);
    rb_define_method(mSysrandom, "__random_uint32", Sysrandom_random_uint32, 0);

    rb_define_singleton_method(mSysrandom, "random_bytes", Sysrandom_random_bytes, -1);
    rb_define_method(mSysrandom, "random_bytes", Sysrandom_random_bytes, -1);
}

/**
 *  call-seq:
 *    Sysrandom#__random_uint32 -> Fixnum
 *
 * Generates a random unsigned 32-bit integer using the OS CSPRNG
 *
 */
static VALUE
Sysrandom_random_uint32(VALUE self)
{
    return UINT2NUM(__randombytes_sysrandom());
}

/**
 *  call-seq:
 *    Sysrandom#random_bytes(n=nil) -> String
 *
 * ::random_bytes generates a random binary string via the OS CSPRNG.
 *
 * The argument n specifies how long the resulting string should be.
 *
 * For compatibility with SecureRandom, if n is not specified, 16 is assumed,
 * and if n is 0, an empty string is returned.
 *
 * The resulting string may contain any byte (i.e. "x00" - "xff")
 *
 */
static VALUE
Sysrandom_random_bytes(int argc, VALUE * argv, VALUE self)
{
    VALUE n_obj, str;
    int n;

    if (rb_scan_args(argc, argv, "01", &n_obj) == 1) {
        if(n_obj == Qnil) {
            n = DEFAULT_N_BYTES;
        } else {
            n = NUM2INT(n_obj);

            if(n < 0) {
              rb_raise(rb_eArgError, "negative string size");
            }
        }
    } else {
        n = DEFAULT_N_BYTES;
    }

    if(n > 0) {
        str = rb_str_new(0, n);
        __randombytes_sysrandom_buf(RSTRING_PTR(str), n);
    } else {
        // n == 0, return empty string
        str = rb_str_new(0, 0);
    }

    return str;
}


ext/sysrandom/randombytes_sysrandom.c
/*
 * __randombytes_sysrandom.c: adapted from libsodium
 * Copyright (c) 2013-2016 Frank Denis <j at pureftpd dot org>
 * https://github.com/jedisct1/libsodium
 */

#include <stdlib.h>
#include <sys/types.h>
#ifndef _WIN32
# include <sys/stat.h>
# include <sys/time.h>
#endif
#ifdef __linux__
# include <sys/syscall.h>
#endif

#include <assert.h>
#include <errno.h>
#include <fcntl.h>
#include <limits.h>
#include <stdint.h>
#include <string.h>
#ifndef _WIN32
# include <unistd.h>
#endif

#ifdef _WIN32
# include <windows.h>
# define RtlGenRandom SystemFunction036
# if defined(__cplusplus)
extern "C"
# endif
BOOLEAN NTAPI RtlGenRandom(PVOID RandomBuffer, ULONG RandomBufferLength);
# pragma comment(lib, "advapi32.lib")
#endif

#if defined(__OpenBSD__) || defined(__CloudABI__)
# define HAVE_SAFE_ARC4RANDOM 1
#endif

#ifndef SSIZE_MAX
# define SSIZE_MAX (SIZE_MAX / 2 - 1)
#endif

#ifdef HAVE_SAFE_ARC4RANDOM

uint32_t
__randombytes_sysrandom(void)
{
    return arc4random();
}

static void
__randombytes_sysrandom_stir(void)
{
}

void
__randombytes_sysrandom_buf(void * const buf, const size_t size)
{
    return arc4random_buf(buf, size);
}

#else /* __OpenBSD__ */

typedef struct SysRandom_ {
    int random_data_source_fd;
    int initialized;
    int getrandom_available;
} SysRandom;

static SysRandom stream = {
    .random_data_source_fd = -1,
    .initialized = 0,
    .getrandom_available = 0
};

#ifndef _WIN32
static ssize_t
safe_read(const int fd, void * const buf_, size_t size)
{
    unsigned char *buf = (unsigned char *) buf_;
    ssize_t        readnb;

    assert(size > (size_t) 0U);
    assert(size <= SSIZE_MAX);
    do {
        while ((readnb = read(fd, buf, size)) < (ssize_t) 0 &&
               (errno == EINTR || errno == EAGAIN)); /* LCOV_EXCL_LINE */
        if (readnb < (ssize_t) 0) {
            return readnb; /* LCOV_EXCL_LINE */
        }
        if (readnb == (ssize_t) 0) {
            break; /* LCOV_EXCL_LINE */
        }
        size -= (size_t) readnb;
        buf += readnb;
    } while (size > (ssize_t) 0);

    return (ssize_t) (buf - (unsigned char *) buf_);
}
#endif

#ifndef _WIN32
static int
__randombytes_sysrandom_random_dev_open(void)
{
/* LCOV_EXCL_START */
    struct stat        st;
    static const char *devices[] = {
# ifndef USE_BLOCKING_RANDOM
        "/dev/urandom",
# endif
        "/dev/random", NULL
    };
    const char **      device = devices;
    int                fd;

    do {
        fd = open(*device, O_RDONLY);
        if (fd != -1) {
            if (fstat(fd, &st) == 0 &&
# ifdef __COMPCERT__
                1
# elif defined(S_ISNAM)
                (S_ISNAM(st.st_mode) || S_ISCHR(st.st_mode))
# else
                S_ISCHR(st.st_mode)
# endif
               ) {
# if defined(F_SETFD) && defined(FD_CLOEXEC)
                (void) fcntl(fd, F_SETFD, fcntl(fd, F_GETFD) | FD_CLOEXEC);
# endif
                return fd;
            }
            (void) close(fd);
        } else if (errno == EINTR) {
            continue;
        }
        device++;
    } while (*device != NULL);

    errno = EIO;
    return -1;
/* LCOV_EXCL_STOP */
}

# ifdef SYS_getrandom
static int
_randombytes_linux_getrandom(void * const buf, const size_t size)
{
    int readnb;

    assert(size <= 256U);
    do {
        readnb = syscall(SYS_getrandom, buf, (int) size, 0);
    } while (readnb < 0 && (errno == EINTR || errno == EAGAIN));

    return (readnb == (int) size) - 1;
}

static int
randombytes_linux_getrandom(void * const buf_, size_t size)
{
    unsigned char *buf = (unsigned char *) buf_;
    size_t         chunk_size = 256U;

    do {
        if (size < chunk_size) {
            chunk_size = size;
            assert(chunk_size > (size_t) 0U);
        }
        if (_randombytes_linux_getrandom(buf, chunk_size) != 0) {
            return -1;
        }
        size -= chunk_size;
        buf += chunk_size;
    } while (size > (size_t) 0U);

    return 0;
}
# endif

static void
__randombytes_sysrandom_init(void)
{
    const int     errno_save = errno;

# ifdef SYS_getrandom
    {
        unsigned char fodder[16];

        if (randombytes_linux_getrandom(fodder, sizeof fodder) == 0) {
            stream.getrandom_available = 1;
            errno = errno_save;
            return;
        }
        stream.getrandom_available = 0;
    }
# endif

    if ((stream.random_data_source_fd =
         __randombytes_sysrandom_random_dev_open()) == -1) {
        abort(); /* LCOV_EXCL_LINE */
    }
    errno = errno_save;
}

#else /* _WIN32 */

static void
__randombytes_sysrandom_init(void)
{
}
#endif

static void
__randombytes_sysrandom_stir(void)
{
    if (stream.initialized == 0) {
        __randombytes_sysrandom_init();
        stream.initialized = 1;
    }
}

static void
__randombytes_sysrandom_stir_if_needed(void)
{
    if (stream.initialized == 0) {
        __randombytes_sysrandom_stir();
    }
}

void
__randombytes_sysrandom_buf(void * const buf, const size_t size)
{
    __randombytes_sysrandom_stir_if_needed();
#ifdef ULONG_LONG_MAX
    /* coverity[result_independent_of_operands] */
    assert(size <= ULONG_LONG_MAX);
#endif
#ifndef _WIN32
# ifdef SYS_getrandom
    if (stream.getrandom_available != 0) {
        if (randombytes_linux_getrandom(buf, size) != 0) {
            abort();
        }
        return;
    }
# endif
    if (stream.random_data_source_fd == -1 ||
        safe_read(stream.random_data_source_fd, buf, size) != (ssize_t) size) {
        abort(); /* LCOV_EXCL_LINE */
    }
#else
    if (size > (size_t) 0xffffffff) {
        abort(); /* LCOV_EXCL_LINE */
    }
    if (! RtlGenRandom((PVOID) buf, (ULONG) size)) {
        abort(); /* LCOV_EXCL_LINE */
    }
#endif
}

uint32_t
__randombytes_sysrandom(void)
{
    uint32_t r;

    __randombytes_sysrandom_buf(&r, sizeof r);

    return r;
}

#endif /* __OpenBSD__ */


lib/sysrandom.rb
require "sysrandom/version"

require "base64"

# Secure random number generation using system RNG facilities
module Sysrandom
  module_function

  # For some reason SecureRandom defaults to 16 bytes
  DEFAULT_LENGTH = 16

  if defined?(JRUBY_VERSION)
    require "java"

    begin
      # Try to use the new NativePRNGNonBlocking algorithm introduced in Java 8.
      @_java_secure_random = java.security.SecureRandom.getInstance("NativePRNGNonBlocking")
    rescue
      # If unavailable, fall back to the default configuration. This will
      # probably be SHA1PRNG, but it depends on the JRE's configuration.
      @_java_secure_random = java.security.SecureRandom.new
    end

    # Random uint32, used by random_number. The C extension provides an equivalent method
    def __random_uint32
      @_java_secure_random.nextLong & 0xFFFFFFFF
    end

    def random_bytes(n = nil)
      n ||= DEFAULT_LENGTH
      raise ArgumentError, "negative string size" if n < 0
      return "" if n == 0

      bytes = Java::byte[n].new
      @_java_secure_random.nextBytes(bytes)
      String.from_java_bytes(bytes)
    end
  else
    require "sysrandom_ext"
  end

  def random_number(n = 0)
    result = __random_uint32 / (2**32).to_f

    if n <= 0
      result
    else
      result *= n
      n.is_a?(Fixnum) ? result.floor : result
    end
  end

  def base64(n = nil)
    Base64.encode64(random_bytes(n)).chomp
  end

  def urlsafe_base64(n = nil, padding = false)
    result = Base64.urlsafe_encode64(random_bytes(n)).chomp
    padding ? result : result.tr("=", "")
  end

  def hex(n = nil)
    random_bytes(n).unpack("h*").first
  end

  def uuid
    values = hex(16).match(/\A(.{8})(.{4})(.)(.{3})(.)(.{3})(.{12})\z/)
    "#{values[1]}-#{values[2]}-4#{values[4]}-#{'89ab'[values[5].ord % 4]}#{values[6]}-#{values[7]}"
  end
end
